File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.py
def solution(board, moves):
    answer = 0
    x, y = len(board), len(board[0])
    st = [[] for _ in range(y)]
    a=0
    for i in range(x-1,-1,-1):
        for j in range(y):
            if board[i][j] > 0: 
                st[j].append(board[i][j])
        a += 1
    
    basket = []
    for k in moves:
        if len(st[k-1])>0:
            doll = st[k-1].pop()
            if len(basket) >0 and basket[-1] == doll:
                basket.pop()
                answer+=2
            else:
                basket.append(doll)

    return answer

File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.java
import java.util.*;
class Solution {
    public int solution(int[][] board, int[] moves) {
        int answer = 0;
        int x = board.length;
        int y = board[0].length;

        Stack<Integer>[] st = new Stack[x];
        for(int k=0; k<y; k++){
            st[k] = new Stack<>();
        }

        for(int i=x-1; i>-1; i--){
            for(int j=0; j<y; j++){
                if (board[i][j]>0)
                    st[j].push(board[i][j]);
            }
        }

        Stack<Integer> basket = new Stack<>();

        for(int i:moves){
            if(!st[i-1].isEmpty()){
                int doll = st[i-1].pop();
                if(!basket.isEmpty() && basket.peek() == doll){
                    basket.pop();
                    answer += 2;
                }else{
                    basket.push(doll);
                }
            }
        }

        return answer;
    }
}